components/Banner/src/Theming/MDCBannerView+MaterialTheming.m

```diff
@@ -0,0 +1,66 @@
+// Copyright 2019-present the Material Components for iOS authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import "MDCBannerView+MaterialTheming.h"
+
+#import <MaterialComponents/MaterialButtons+Theming.h>
+#import <MaterialComponents/MaterialButtons.h>
+
+// The opacity value applied to text view by default.
+static CGFloat const kTextViewOpacityDefault = (CGFloat)0.87;
+// The opacity value applied to divider by default.
+static CGFloat const kDividerOpacityDefault = (CGFloat)0.12;
+
+@implementation MDCBannerView (MaterialTheming)
+
+- (void)applyThemeWithScheme:(nonnull id<MDCContainerScheming>)scheme {
+  // Color
+  [self applyThemeWithColorScheme:scheme.colorScheme];
+
+  // Typography
+  [self applyThemeWithTypographyScheme:scheme.typographyScheme];
+
+  // Apply Text Theme to buttons on Banner by default.
+  [self.leadingButton applyTextThemeWithScheme:scheme];
+  [self.trailingButton applyTextThemeWithScheme:scheme];
+
+  self.showsDivider = YES;
+}
+
+#pragma mark - Internal Helpers
+
+- (void)applyThemeWithColorScheme:(id<MDCColorScheming>)colorScheme {
+  self.backgroundColor = colorScheme.surfaceColor;
+  self.textView.textColor =
+      [colorScheme.onSurfaceColor colorWithAlphaComponent:kTextViewOpacityDefault];
+  self.dividerColor = [colorScheme.onSurfaceColor colorWithAlphaComponent:kDividerOpacityDefault];
+}
+
+- (void)applyThemeWithTypographyScheme:(id<MDCTypographyScheming>)typographyScheme {
+  UIFont *textFont = typographyScheme.body2;
+  if ([typographyScheme respondsToSelector:@selector(useCurrentContentSizeCategoryWhenApplied)]) {
```

Comment:
I'm afraid I don't understand this logic. If the implementation supports both properties, but the client only sets one of them (perhaps they set the `mdc_adjustsFontForContentSizeCategory` because that existed first), their expected behavior will fail, correct?

This would be the case for a client that used a custom implementation of `MDCTypographyScheming` that was written before `useCurrentContentSizeCategoryWhenApplied` was added but now conforms to the optional property. Their configuration code may still be using `mdc_adjustsFontForContentSizeCategory` and not referencing `useCurrentContentSizeCategoryWhenApplied` at all. However, since the class responds to the `useCurrentContentSizeCategoryWhenApplied` selector, this code will never evaluate their expected behavior from `mdc_adjustsFontForContentSizeCategory`.

Since the protocol does not require that these properties share backing data, they could easily have different values (`useCurrentContentSizeCategoryWhenApplied` defaults to `NO` but `mdc_adjustsFontForContentSizeCategory` is set to `YES` by the client.

Reply:
This case could happen, but I don't see why clients would implement optional `useCurrentContentSizeCategoryWhenApplied` property without understanding what it means. It only happens when they choose to use this property, which is why this property is optional for compatibility.

Reply:
Sorry, I wasn't clear that the client and the scheme implementation aren't necessarily the same project. 

*   App `A` uses scheme from Library `L` which conforms to `MDCTypographyScheming`. 
*   `L` updates its scheme to conform to the optional property (a non-breaking, minor update). 
*   `A` is still only setting the legacy API since it might not have even noticed the update (since it requires no code migration), but if they were to call our Theming Extension, they would not receive the correctly-scaled fonts.

I believe the problem lies in the `MDCTypographyScheming` protocol, since it doesn't require that implementations bind the two properties to the same value. As a result, the Theming Extensions now have to work-around this possibility.  At least until we can either remove the old API or update the documentation for them.

Reply:
I don't think this case should be covered by us. If `L` updates their scheme class, they need to understand it would not be fully compatible with our extension based on scheme. Our extension are not able to be aware of those downstream changes. Since we added ` @note This flag will become required in the future as a replacement for
 @c mdc_adjustsFontForContentSizeCategory.` in the protocol. I feel it should be enough? Otherwise, there is no way to handle this case. Because `mdc_adjustsFontForContentSizeCategory` and `useCurrentContentSizeCategoryWhenApplied` have the same semantic meaning, users who use/implement it needs to make sure the value are the same. Maybe adding another note in the `MDCTypographyScheming` to make it clear?

Reply:
I agree that changing the expected contract (documentation) between the two properties is probably the best route forward, since having to accommodate the disparity in every component would become expensive an error-prone. If you'd like to push that up as a PR, I'd be happy to approve it.

Reply:
https://github.com/material-components/material-components-ios/pull/8311 has been created to address it.